src/platform/ESP32/NetworkCommissioningWiFiDriver.cpp

```diff
@@ -0,0 +1,319 @@
+}
+
+void ESPWiFiDriver::OnConnectWiFiNetwork()
+{
+    if (mpConnectCallback)
+    {
+        mpConnectCallback->OnResult(Status::kSuccess, CharSpan(), 0);
+        mpConnectCallback = nullptr;
+    }
+}
+
+void ESPWiFiDriver::ConnectNetwork(ByteSpan networkId, ConnectCallback * callback)
+{
+    CHIP_ERROR err          = CHIP_NO_ERROR;
+    Status networkingStatus = Status::kSuccess;
+
+    VerifyOrExit(NetworkMatch(mStagingNetwork, networkId), networkingStatus = Status::kNetworkIDNotFound);
+    VerifyOrExit(mpConnectCallback == nullptr, networkingStatus = Status::kUnknownError);
+    ChipLogProgress(NetworkProvisioning, "ESP NetworkCommissioningDelegate: SSID: %s", networkId.data());
+
+    err               = ConnectWiFiNetwork(reinterpret_cast<const char *>(mStagingNetwork.ssid), mStagingNetwork.ssidLen,
+                             reinterpret_cast<const char *>(mStagingNetwork.credentials), mStagingNetwork.credentialsLen);
+    mpConnectCallback = callback;
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        networkingStatus = Status::kUnknownError;
+    }
+    if (networkingStatus != Status::kSuccess)
+    {
+        ChipLogError(NetworkProvisioning, "Failed to connect to WiFi network:%s", chip::ErrorStr(err));
+        mpConnectCallback = nullptr;
+        callback->OnResult(networkingStatus, CharSpan(), 0);
+    }
+}
+
+CHIP_ERROR ESPWiFiDriver::StartScanWiFiNetworks(ByteSpan ssid)
+{
+    esp_err_t err = ESP_OK;
+    if (ssid.data())
+    {
+        wifi_scan_config_t scan_config = { 0 };
+        memset(WiFiSSIDStr, 0, sizeof(WiFiSSIDStr));
+        memcpy(WiFiSSIDStr, ssid.data(), ssid.size());
+        scan_config.ssid = WiFiSSIDStr;
+        err              = esp_wifi_scan_start(&scan_config, false);
+    }
+    else
+    {
+        err = esp_wifi_scan_start(NULL, false);
+    }
+    if (err != ESP_OK)
+    {
+        return chip::DeviceLayer::Internal::ESP32Utils::MapError(err);
+    }
+    return CHIP_NO_ERROR;
+}
+
+void ESPWiFiDriver::OnScanWiFiNetworkDone()
+{
+    uint16_t ap_number;
+    esp_wifi_scan_get_ap_num(&ap_number);
+    if (!ap_number)
+    {
+        ChipLogProgress(DeviceLayer, "No AP found");
+        if (mpScanCallback != nullptr)
+        {
+            mpScanCallback->OnFinished(Status::kSuccess, CharSpan(), nullptr);
+            mpScanCallback = nullptr;
+        }
+        return;
+    }
+    wifi_ap_record_t * ap_list_buffer;
+    ap_list_buffer = (wifi_ap_record_t *) malloc(ap_number * sizeof(wifi_ap_record_t));
+    if (ap_list_buffer == NULL)
+    {
+        ChipLogError(DeviceLayer, "can't malloc memory for ap_list_buffer");
+        if (mpScanCallback)
+        {
+            mpScanCallback->OnFinished(Status::kUnknownError, CharSpan(), nullptr);
+            mpScanCallback = nullptr;
+        }
+        return;
+    }
+    if (esp_wifi_scan_get_ap_records(&ap_number, (wifi_ap_record_t *) ap_list_buffer) == ESP_OK)
```

Comment:
Nit: `ap_list_buffer` is already of type `wifi_ap_record_t *`. No need to typecast?

Reply:
Removed it.